### PR TITLE
Added optional function injection

### PR DESCRIPTION
**Pull request by RyRy79261**

### CODE REVIEW

File: src/context/Web3Context.tsx [hunk 1]
```diff
@@ -17,11 +17,14 @@ export type OnboardConfig = Partial<Omit<Initialization, 'networkId'>>;
...
 type EtherchainGasSettings = 'safeLow' | 'standard' | 'fast' | 'fastest';
 
+
+
 type TokenConfig = {
   address: string;
   name?: string;
   symbol?: string;
   imageUri?: string;
+  attachFunctions?: boolean
```
Comment (FSM1) [inline src/context/Web3Context.tsx:27]: what is the reasoning behind this parameter being passed in?

File: src/context/Web3Context.tsx [hunk 2]
```diff
@@ -260,6 +264,14 @@ const Web3Provider = ({
           );
         }
 
+        if (token.attachFunctions) {
+          newTokenInfo.functions = {
+            allowance: tokenContract.allowance,
```
Comment (FSM1) [inline src/context/Web3Context.tsx:269]: These functions should be wrapped in a `useCallback()` and exposed directly on the the `TokenInfo` object on line 234 
```
        const newTokenInfo: TokenInfo = {
          decimals: 0,
          balance: 0,
          name: token.name,
          symbol: token.symbol,
          imageUri: token.imageUri,
          allowance: useCallback()
          ....
        };
```

File: src/context/tokensReducer.ts [hunk 3]
```diff
@@ -1,10 +1,33 @@
+import { BigNumber, BigNumberish, CallOverrides, ContractTransaction, Overrides } from "ethers";
+
+export type TokenFunctions = {
```
Comment (FSM1) [inline src/context/tokensReducer.ts:3]: no need for this wrapper type, just expose these directly in the TokenInfo type

**Review by FSM1 — CHANGES_REQUESTED** [review]

File: src/context/Web3Context.tsx (hunk 1 above)
Comment (RyRy79261) [reply src/context/Web3Context.tsx:27]: In some instances there might not be either a need for passing functions through or that the context could be used without there being a User controlled wallet provider fetching the token data

Comment (FSM1) [reply src/context/Web3Context.tsx:27]: unnecessary overcomplication... if the erc address is being passed in the functions should always be available

**Review by FSM1 — APPROVED** [review]

**Review by FSM1 — APPROVED** [review]

**Pull request merged**